### ISSUE TITLE
XXV710 (i40e) FW 09.05.07: STL latency streams show 100% RX ierrors and 0 latency stats while data streams have 0% loss

### ISSUE THREAD

**Issue by pradiptapks** [perftool-incubator/bench-trafficgen#138]
## Summary

On Intel XXV710 (i40e) with TRex v3.08, STL latency streams are transmitted but never received as latency stats.

- Port `ipackets` match the data stream only.
- Port `ierrors` match the latency TX count exactly (100% latency loss).
- The DUT forwards both stream types with zero drops.

TRex logs `A workaround is applied for issue #508` on every port and `FW ver 09.05.07`. That is the post-7.0 X710/XXV710 firmware class. The published #508 and #496 symptoms (watchdog, no RX, ARP failure) are not what we see here. Data RX is clean. Latency frames come back as NIC RX errors.

Binary search then fails at every rate because `--use-device-stats` folds the constant 1000 pps latency gap into directional loss. Lowering the data rate only increases the reported loss percentage.

Jira Card: [OSPRH-35070](https://redhat.atlassian.net/browse/OSPRH-35070)

## Environment

| Item | Value |
| --- | --- |
| TRex | v3.08 (stateless / trex-txrx) |
| TRex host | nfv-intel-5.xxx.redhat.com |
| NICs | 8x Ethernet Controller XXV710 for 25GbE SFP28 (158b), vfio-pci / net_i40e |
| PCI | 0000:3b:00.[0-1], 0000:5e:00.[0-1], 0000:86:00.[0-1], 0000:af:00.[0-1] |
| Firmware (TRex/DPDK print) | FW ver 09.05.07 on all 8 ports |
| Queue mode | DROP_QUE_FILTER (rx_qid 0 drop/64 desc, rx_qid 1 data/4096 desc) |
| Cores | c: 6 (not --software -c 1) |
| Checksum offload | on (TRex default) |
| 1588 | ports report is_ieee1588_supported: yes, rx_filter_mode: hardware match |
| DDP | Loading DDP profile ... Profile is already loaded |
| Issue #508 | A workaround is applied for issue #508 on all 8 ports |
| DUT | testpmd io, DPDK 25.11, mlx5 VM (tenant-sriov-vm1), 2 ports, 0 RX/TX errors, 0 dropped |
| Orchestration | Crucible / trafficgen binary-search.py + trex-txrx.py |
| Run | [trafficgen--2026-08-19_05:59:14_UTC--94e69661-5e7d-40fa-9a21-7170b054d406](https://storage.scalelab.redhat.com/psahoo/PerfTaskLog/Crucible_RUN/trafficgen--2026-08-19_05%3A59%3A14_UTC--94e69661-5e7d-40fa-9a21-7170b054d406.tar.xz) |

09.05.07 is the DPDK AQ firmware triple, not the Intel marketing NVM string (9.40 / 9.52 / ...). Host NVM / eetrack still needs `ethtool -i` and `dmesg | grep nvm`.

## TRex config

    c: 6
    interfaces:
      - 0000:3b:00.0
      - 0000:3b:00.1
      - 0000:5e:00.0
      - 0000:5e:00.1
      - 0000:86:00.0
      - 0000:86:00.1
      - 0000:af:00.0
      - 0000:af:00.1
    port_bandwidth_gb: 25

Traffic:

- Direction: bidirectional, 4 pairs 0:1,2:3,4:5,6:7
- Data (default): UDP, 512B, stream_mode=burst, ~58.7 kpps/port at 1% line rate
- Latency: UDP, 512B, STLFlowLatencyStats, 1000 pps, stream_mode=continuous
- --skip-hw-flow-stats for data; latency still uses HW flow/latency stats
- --use-device-stats for pass/fail loss
- max-loss-pct=0.002

Log excerpts:

    set driver name net_i40e
    set dpdk queues mode to DROP_QUE_FILTER
    port 0: FW ver 09.05.07
    A workaround is applied for issue #508
    Warning --checksum-offload is turn on by default, no need to call it

## Observed counters

Trial 2 (search, 1%, 90s). Same split on all 8 ports:

| Counter | Per port | Meaning |
| --- | ---: | --- |
| TX total | 5,376,654 | data + latency |
| TX latency (pg_id 16/47/78/109/140/171/202/233) | 90,000 | 1000 pps x 90s |
| TX data (implied) | 5,286,654 | TX minus latency |
| RX ipackets | 5,286,654 | exactly data TX |
| RX latency (flow_stats[].rx_pkts.total) | 0 | 100% latency loss |
| ierrors | 90,000 | exactly latency TX |
| Latency average | -1.0 | no samples |
| Latency err_cntrs / bad_hdr | 0 | software RX never saw the frames |

Trial 9 (0.01%) is the same pattern: data TX 52,866 = RX 52,866; latency TX 90,001 / RX 0; ierrors 90,001.

Binary-search log on every pair, every trial:

    (trial information, individual stream 100% RX packet loss , device pair: 0 -> 1, pg_ids: [16])
    ... same for pg_ids [47], [78], [109], [140], [171], [202], [233]
    (trial failed requirement, percent loss, direction: ->, requested: 0.002%, achieved: 1.67%, lost packets: 360,000)

Lost packets per direction = 4 pairs x 90,000 latency = 360,000. All reported loss is latency.

## DUT is not dropping

testpmd io, mlx5, 1 queue/port, RX offloads 0x0:

- Port 0: RX 52,260,108 = TX 52,260,108, dropped 0, RX-errors 0
- Port 1: same
- Combined 104,520,216 matches the sum of TRex TX across all trials plus teaching warmup

Latency packets hairpin through the DUT. TRex i40e classifies them as ierrors / zero latency stats.

## Why search cannot pass

Latency is a fixed 1000 pps. Device-stat loss = latency TX / (data TX + latency TX):

| Rate | Directional loss |
| --- | ---: |
| 1.000% | 1.67% |
| 0.500% | 3.29% |
| 0.125% | 11.98% |
| 0.010% | 63.00% |

Halving the data rate only makes latency a larger fraction of the mix. With max-loss-pct 0.002 there is no passing rate while latency RX stays at 0.

## Related upstream

These are the closest public reports. None is a 1:1 match to ierrors == latency TX with perfect data ipackets.

1. https://github.com/cisco-system-traffic-generator/trex-core/issues/508
   X?710 NVM 7.0.x+: LL TX QoS firmware command fails. Workaround in v2.82+ (dummy TX queue; LL QoS off). Quote: "latency packets resolution will be impacted." This run applies that workaround. It does not by itself explain 100% RX ierrors.

2. https://github.com/cisco-system-traffic-generator/trex-core/issues/496
   XXV710 NVM 7.10 vs working 6.00: --software -c 1 receives nothing / ARP fails. Closed as the same firmware class as #508. Not this config (c: 6, data RX OK).

3. https://github.com/cisco-system-traffic-generator/trex-core/issues/384
   https://groups.google.com/g/trex-tgn/c/NepxT6j8on4
   NVM 6.0.1 drop-queue 27 Mpps by design; downgrade to 5.0.4/5.0.5 or DDP. Rate here is ~0.5 Mpps; DDP already loaded.

4. https://github.com/cisco-system-traffic-generator/trex-core/issues/631
   https://github.com/cisco-system-traffic-generator/trex-core/issues/882
   DROP_QUE_FILTER HW filter miss (IP ID FLOW_STAT_PAYLOAD_IP_ID / TOS LSB) leads to latency RX 0. Those reports usually still increment ipackets and leave ierrors=0. That is the opposite of the ierrors split seen here.

5. Intel X710/XXV710/XL710 Specification Update: https://cdrdv2.intel.com/v1/dl/getContent/331430
   - Errata 76 (NoFix): incorrect FDIR flex-payload extraction to RX descriptor
   - Errata 89 (NoFix): IP packets in low-latency TC not fully processed (no IP/L4 checksum validation)
   - Spec change 20: IEEE 1588 over UDP not processed in NVM 5.05 (L2 PTP only); UDP PTP restored in NVM 6.01

NVM mapping: https://cdrdv2.intel.com/v1/dl/getContent/332191
TRex NVM notes: https://trex-tgn.cisco.com/trex/doc/trex_manual.html#_firmware_update_to_xl710_x710

## Hypotheses

1. FDIR / drop-queue on FW 9.x: latency frames are not steered to rx_qid 1 and are counted as ierrors rather than good ipackets (different from #631).
2. RX checksum offload: latency streams rewrite payload (seq/timestamp) every packet; a stale UDP checksum causes i40e to flag only latency frames. Data streams are unmodified and pass.
3. PTP / 1588 parse on 9.x NVM: ports advertise 1588; the latency header/UDP may be treated as TimeSync and errored (Intel spec change 20 / TRex PR #712).
4. #508 dummy LL queue: known to hurt latency resolution; worth confirming it does not also break latency receive classification on NVM 9.x.

## What this is not

- DUT capacity / testpmd drops (0 dropped, RX=TX)
- Line-rate / 27 Mpps drop-queue cap (#384)
- --software -c 1 total RX blackhole (#496)
- Data-stream loss (0% at every rate including 1%)

## Questions

1. Is ierrors == latency TX with ipackets == data TX a known i40e / FW 9.x behavior in DROP_QUE_FILTER?
2. After the #508 workaround, should STL latency RX still work on XXV710 NVM 9.x with checksum offload on?
3. Recommended isolation:
   - --software (bypass FDIR)
   - disable RX checksum offload
   - measure-latency=0 (confirm data-only pass)
   - capture one bounced latency frame (UDP checksum, IPv4 ID, TOS LSB, PTP-like header)

Happy to attach binary-search.trial-002.extra.txt (parsable port stats + per-pg_id flow/latency) and the TRex server log around FW / #508.

## Test plan

- [ ] XXV710, TRex >= 3.08, NVM 9.x, DROP_QUE_FILTER, STL data + latency, 1000 pps latency, bidirectional
- [ ] Latency flow_stats RX > 0 and ierrors not equal to latency TX
- [ ] Same with checksum offload on (TRex default)
- [ ] Same after the #508 workaround is applied
- [ ] DUT io-forwarder (or cable loopback) does not rewrite IPv4 ID / TOS

**Comment by pradiptapks** [comment]
@atheurer @k-rister Follow-up analysis is in the below (switch loopback, NVM locked on 9.5.x, trex-core#1222).

**Throughput on these XXV710s works** with `measure-latency=0`. **TRex HW latency does not** (`DROP_QUE_FILTER` + #508 workaround; latency frames come back as `ierrors`). We cannot downgrade below 9.x (EEPID / MinSRev), and 9.50 vs 9.57 does not change the result. `--software` makes RX worse, so that is not a fallback.

Please advise on the ticket:

1. Document `measure-latency=0` for XXV710 / i40e NVM 9.x and **close this as external** (blocked on trex-core#1222), or
2. **Keep it open** until TRex responds.

To keep benchmarking this hardware, I plan to leave TRex STL backend with (`measure-latency=0`) and collect latency with **`ptp-latency`** (kernel `SO_TIMESTAMPING`). I already validated that path successfully on an additional XXV710. 

Does that match for nfv benchmark standard, could you please share your recommendation on this?

## Follow-up : isolation complete

This is **not a trafficgen/binary-search bug**. Data forwarding is clean. STL **latency frames are TX'd and come back as i40e `ierrors`**, so `STLFlowLatencyStats` RX stays 0. `--use-device-stats` then reports that gap as directional loss.

TRex-native tracker: [cisco-system-traffic-generator/trex-core#1222](https://github.com/cisco-system-traffic-generator/trex-core/issues/1222)

### Experiments since the original report

| Change | Result |
|---|---|
| DUT in path (baseline) | Fail: data OK, latency 100% `ierrors` |
| `trex-software-mode=on` / `--software` | **Worse.** Data RX also dies (#496 class). Do not retry on this STL/XXV710/v3.08 combo. |
| `--checksum-offload-disable` | **No change.** TX checksum offload is not the cause. |
| `measure-latency=0` | **Pass.** 0 `ierrors`. Data path is fine. |
| Switch loopback, no DUT | **Same fail.** DUT/OSP are out of the path. |
| FW `09.05.00` (still 9.5.x) | Identical fail / identical counters. |
| Intel NVM **9.57** (TRex still prints `FW ver 09.05.07`) | Identical fail / identical counters. |

Loopback validation, 120s @ 1%, port 0, **all three 9.5.x images**:

`opackets` 7,168,872 / `ipackets` 7,048,872 / `ierrors` **120,000** / latency RX **0**

`ierrors == latency TX` (1000 pps × 120s). Directional “loss” 1.67% is only those latency frames.

### Firmware: cannot leave 9.x

Host NVM is **9.57 `0x800103a8`** (Device ID `158B`). A 7.10 `nvmupdate64e -u` run on `-location 00:059` (`0000:3b:00.0`) **did not write NVM**.

- Result: `No devices to update` / XML `Fail id="8"` — nothing eligible, not a bad location.
- EEPID `800103A8` is not in that kit’s `REPLACES` (kit only knows older `8000Dxxx`).
- MinSRev does not go backwards: 9.57 is SRev **40**, 7.10 is SRev **10**. Silicon would reject 7.10 even if forced.
- Intel’s current 9.57 downgrade kits only go **9.57 → 9.56 … 9.50**. The old 9.30 → 7.10 path does not apply.

Do **not** add `800103A8` to `REPLACES` or force a `.bin` (recover-lock risk). All six cards report the same EEPID; repeating `-u` on the other locations with the 7.10 package will also skip.

TRex only skips the #508 dummy TX queue if packed FW is **≤ 6.01**. That NVM is unpublished and **unreachable** on this silicon. 9.50 vs 9.57 does not change TRex’s `version > 0x6001` gate.

### Original hypotheses

| # | Hypothesis | Status |
|---|---|---|
| 1 | FDIR / drop-queue on FW 9.x: latency frames not steered to `rx_qid` 1, counted as `ierrors` | **Still the best fit.** Confirmed with loopback. Packets never reach SW (`bad_hdr` 0). |
| 2 | TX/RX checksum offload on rewritten latency payload | **Not TX checksum.** `--checksum-offload-disable` unchanged. RX checksum was not toggled independently; low priority given the exact `ierrors == latency TX` split. |
| 3 | IEEE 1588 / PTP parse | **Unlikely.** Default v3.08 does not enable `RTE_LIBRTE_IEEE1588` (no `TIMESYNC` in logs). PR #712 is compile-time accuracy for PTP packets, not a fix for this `ierrors` split. |
| 4 | #508 dummy LL TX queue | **Always applied** on these cards (`FW ver 09.05.0x` > `0x6001`). That workaround was for watchdog / missing LL QoS, not 100% RX `ierrors`. It is still on and latency RX is still 0. |

### trex-core code review (no in-tree fix)

- Latency: IP ID `0xFFFF` + TOS LSB + 16-byte magic → FDIR to `MAIN_DPDK_RX_Q`. Data/flow-stat: reserved IP ID → drop queue; NIC still counts `ipackets`.
- #508: `extra_tx_queues_requires()` adds a dummy TX queue. Already on. Open PRs do not touch i40e latency `ierrors`.
- `--software` is the documented HW-filter escape and is **unusable** here (#496).
- `--no-hw-flow-stat` only changes flow-stat counting; latency still needs FDIR → RX_Q.

MAC `ierrors` happen **before** DPDK delivery. trafficgen cannot count those frames.

### Practical workaround (trafficgen)

On XXV710 NVM 9.5.x + TRex v3.08 STL:

- **Throughput:** `measure-latency=0`. Proven pass.
- **Latency numbers:** use a NIC TRex still measures in HW (E810/mlx5), or skip HW latency on this host.
- Do **not** retry `--software`, checksum offload, or NVM &lt; 9.50.

Optional remaining evidence for trex-core #1222 (not required to close this trafficgen ticket): SPAN/capture one data frame vs one latency frame on the loopback. If latency looks valid on the wire → RX FDIR/DDP. If truncated/bad checksum → #508 TX-queue path.

### Proposed trafficgen follow-ups

1. Document: XXV710 / i40e NVM 9.x + `measure-latency=1` in `DROP_QUE_FILTER` is a known TRex failure; use `measure-latency=0`.
2. Keep this issue open against trex-core #1222, or close as **wontfix / external** with that link and the workaround above.
